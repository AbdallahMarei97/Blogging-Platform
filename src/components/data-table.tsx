"use client";

import { ReactNode } from "react";
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableHeader,
  TableRow,
} from "@/components/ui/table";
import { cn } from "@/lib/utils";

export interface Column<TData> {
  accessorKey?: string;
  id: string;
  className?: string;
  header: string;
  cell?: (item: TData) => ReactNode;
}

export function DataTable<TData>({
  data,
  columns,
}: {
  data: TData[];
  columns: Column<TData>[];
}) {
  return (
    <Table>
      <TableHeader>
        <TableRow>
          {columns.map((column) => (
            <TableHead
              key={column.accessorKey || column.id}
              className={cn(column.className, "text-center")}
            >
              {column.header}
            </TableHead>
          ))}
        </TableRow>
      </TableHeader>
      <TableBody>
        {data?.map((item, index) => (
          <TableRow key={index}>
            {columns.map((column) => (
              <TableCell
                key={column.accessorKey || column.id}
                className={cn(column.className, "text-center")}
              >
                {column.cell?.(item)}
              </TableCell>
            ))}
          </TableRow>
        ))}
      </TableBody>
    </Table>
  );
}
