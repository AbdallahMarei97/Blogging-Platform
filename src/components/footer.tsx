export const Footer: React.FC = () => {
  return (
    <footer className="bg-black text-white py-4 mt-9">
      <div className="container mx-auto text-center">
        <p className="text-sm">
          Â© {new Date().getFullYear()} Blogging Platform. All rights reserved.
        </p>
        <p className="text-sm mt-2">
          Built by{" "}
          <a href="https://github.com/AbdallahMarei97" className="underline">
            Abdallah Marei
          </a>
        </p>
      </div>
    </footer>
  );
};
