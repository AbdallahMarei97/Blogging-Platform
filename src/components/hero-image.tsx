interface HeroImageProps {
  title: string;
  subtitle: string;
}

export const HeroImage: React.FC<HeroImageProps> = ({ title, subtitle }) => {
  return (
    <div className="relative w-full h-64 md:h-96 mb-3">
      <div className="absolute inset-0 bg-black bg-opacity-50 flex flex-col justify-center items-center text-center text-white p-4">
        <h1 className="text-3xl md:text-5xl font-bold">{title}</h1>
        <p className="text-lg md:text-2xl mt-2">{subtitle}</p>
      </div>
    </div>
  );
};
