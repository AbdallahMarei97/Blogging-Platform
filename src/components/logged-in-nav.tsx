"use client";
import Link from "next/link";
import { signOut } from "next-auth/react";

import { LogOut, LayoutDashboard } from "lucide-react";
import { Avatar, AvatarFallback } from "@/components/ui/avatar";
import { Button } from "@/components/ui/button";
import {
  DropdownMenu,
  DropdownMenuContent,
  DropdownMenuGroup,
  DropdownMenuItem,
  DropdownMenuLabel,
  DropdownMenuSeparator,
  DropdownMenuTrigger,
} from "@/components/ui/dropdown-menu";

export const LoggedInNav = ({
  name,
  email,
}: {
  name: string | null | undefined;
  email: string | null | undefined;
}) => {
  return (
    <DropdownMenu>
      <DropdownMenuTrigger asChild>
        <Button variant="ghost" className="relative h-8 w-8 rounded-full">
          <Avatar className="h-8 w-8">
            <AvatarFallback>{name?.slice(0, 2)}</AvatarFallback>
          </Avatar>
        </Button>
      </DropdownMenuTrigger>
      <DropdownMenuContent className="w-56" align="end" forceMount>
        <DropdownMenuLabel className="font-normal">
          <div className="flex flex-col space-y-1">
            <p className="line-clamp-2 text-sm font-medium leading-none">
              {name}
            </p>
            <p className="truncate text-xs leading-none text-muted-foreground">
              {email}
            </p>
          </div>
        </DropdownMenuLabel>
        <DropdownMenuGroup>
          <DropdownMenuItem>
            <Link href="/dashboard/" className="flex w-full items-center gap-2">
              <LayoutDashboard className="h-4 w-4" />
              Dashboard
            </Link>
          </DropdownMenuItem>
        </DropdownMenuGroup>
        <DropdownMenuSeparator />
        <DropdownMenuItem
          className="cursor-pointer"
          onClick={() => signOut({ redirectTo: "/login" })}
        >
          <button className="flex items-center gap-2">
            <LogOut className="h-4 w-4" />
            <span>Log out</span>
          </button>
        </DropdownMenuItem>
      </DropdownMenuContent>
    </DropdownMenu>
  );
};
