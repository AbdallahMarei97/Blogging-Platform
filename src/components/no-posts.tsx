import Link from "next/link";
import { Button } from "@/components/ui/button";
import { auth } from "@/auth";

export default async function NoPosts() {
  const session = await auth();
  return (
    <>
      <div className="container text-center mt-9">
        <div className="space-y-2">
          <h1 className="text-3xl font-bold tracking-tight sm:text-5xl">
            No Posts
          </h1>
          <p className="mb-4 text-lg text-gray-700">
            Sorry, there are no posts created yet.
          </p>
          <Button asChild>
            {session?.user ? (
              <Link href="/dashboard">
                Go to your dashboard to create posts
              </Link>
            ) : (
              <Link href="/login">Login to start creating posts</Link>
            )}
          </Button>
        </div>
      </div>
    </>
  );
}
