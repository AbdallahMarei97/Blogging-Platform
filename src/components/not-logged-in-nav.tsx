import Link from "next/link";

export const NotLoggedInNav = () => {
  return (
    <nav className="flex items-center gap-1">
      <Link
        className="text-sm font-medium text-foreground/80 transition-colors hover:text-foreground"
        href="/login/"
      >
        Login
      </Link>
      /
      <Link
        className="text-sm font-medium text-foreground/80 transition-colors hover:text-foreground"
        href="/sign-up/"
      >
        Sign Up
      </Link>
    </nav>
  );
};
