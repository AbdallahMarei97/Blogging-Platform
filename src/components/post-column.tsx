"use client";
import { deletePost } from "@/actions/post";
import { type Column } from "./data-table";
import { type post } from "@/db/schema";
import { InferSelectModel } from "drizzle-orm";
import { DeletePost } from "./post/delete-post";
import UpdatePost from "./post/update-post";

type Post = InferSelectModel<typeof post>;

export const postColumns: Column<Post>[] = [
  {
    id: "title",
    header: "Title",
    accessorKey: "title",
    className: "w-30 whitespace-normal",
    cell: (item) => item.title,
  },
  {
    id: "body",
    header: "Body",
    accessorKey: "body",
    className: "max-w-46 whitespace-normal break-words",
    cell: (item) => item.body,
  },
  {
    id: "createdAt",
    header: "Created At",
    accessorKey: "createdAt",
    className: "w-20",
    cell: (item) => new Date(item.createdAt).toLocaleDateString(),
  },
  {
    id: "action",
    header: "Action",
    className: "w-36",
    cell: (item) => (
      <div className="flex justify-center space-x-2">
        <UpdatePost {...item} />
        <DeletePost onConfirm={() => deletePost(item.id, item.userId)} />
      </div>
    ),
  },
];
