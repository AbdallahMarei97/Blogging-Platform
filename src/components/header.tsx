import React from "react";
import { NotLoggedInNav } from "./not-logged-in-nav";
import { LoggedInNav } from "./logged-in-nav";
import { auth } from "@/auth";
import Link from "next/link";

const Header = async () => {
  const session = await auth();
  return (
    <div className="supports-backdrop-blur:bg-background/60 sticky left-0 right-0 top-0 z-20 border-b bg-background/95 px-4 backdrop-blur">
      <div className="flex h-14 items-center justify-between gap-2">
        <Link
          href="/"
          className="flex items-center gap-2 text-xl font-bold tracking-tighter md:text-2xl"
        >
          Blogging Platform
        </Link>
        <div className="flex items-center gap-2">
          {!session?.user ? (
            <NotLoggedInNav />
          ) : (
            <LoggedInNav
              name={session?.user?.userName}
              email={session?.user?.email}
            />
          )}
        </div>
      </div>
    </div>
  );
};

export default Header;
