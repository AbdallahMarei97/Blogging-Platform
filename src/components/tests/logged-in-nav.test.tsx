import { render, screen } from "@testing-library/react";
import { LoggedInNav } from "@/components/logged-in-nav";

describe("NotLoggedInNav", () => {
  it("should render the Login Button", async () => {
    render(<LoggedInNav name="Abdallah" email="abdallah@test.com" />);
    const LoginButton = await screen.findByText("Ab");
    expect(LoginButton).toBeDefined();
  });

  it("should match snapshot", async () => {
    const { container } = render(
      <LoggedInNav name="Abdallah" email="abdallah@test.com" />
    );
    expect(container).toMatchSnapshot();
  });
});
