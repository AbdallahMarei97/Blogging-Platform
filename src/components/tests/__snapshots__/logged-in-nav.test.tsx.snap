// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotLoggedInNav > should match snapshot 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-all disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg:not([class*='size-'])]:size-4 shrink-0 [&_svg]:shrink-0 outline-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive hover:bg-accent hover:text-accent-foreground dark:hover:bg-accent/50 px-4 py-2 has-[>svg]:px-3 relative h-8 w-8 rounded-full"
    data-slot="dropdown-menu-trigger"
    data-state="closed"
    id="radix-:r2:"
    type="button"
  >
    <span
      class="relative flex size-8 shrink-0 overflow-hidden rounded-full h-8 w-8"
      data-slot="avatar"
    >
      <span
        class="bg-muted flex size-full items-center justify-center rounded-full"
        data-slot="avatar-fallback"
      >
        Ab
      </span>
    </span>
  </button>
</div>
`;

exports[`NotLoggedInNav should match snapshot 1`] = `
<div>
  <nav
    class="flex items-center gap-1"
  >
    <a
      class="text-sm font-medium text-foreground/80 transition-colors hover:text-foreground"
      href="/login"
    >
      Login
    </a>
    /
    <a
      class="text-sm font-medium text-foreground/80 transition-colors hover:text-foreground"
      href="/sign-up"
    >
      Sign Up
    </a>
  </nav>
</div>
`;
